### ISSUE TITLE
Atomic creation of .drv files to Nix Store

### ISSUE THREAD

**Issue by volth** [NixOS/nix#3498]
The most frequent Nix Store corruption caused by sudden hang or power outage is zero-length .drv files.

After that, Nix Store is unusable until `nix-store --verify --repair --check-contents` is performed to find all corruption (which might take hours)
Typically there are only tons of
```
path '/nix/store/kq2gq6vkiqsw7q9gp3j3q5dfkj82s0w0-closure-info.drv' was modified! expected hash 'sha256:02xp9wps6yixn8q2jkvxhrhl3gcv1316v8ll9y5p6cpk4isiciiq', got 'sha256:0ip26j2h11n1kgkz36rl4akv694yz65hr72q4kv4b3lxcbi65b3p' 
path '/nix/store/kpghq1j486cbhfz2d22qvr880mkvypgq-etc.drv' was modified! expected hash 'sha256:0gf2jvp6xlgbsxf7fby1bhm7mkprws8p9djls0ymngc1a3rnylv7', got 'sha256:0ip26j2h11n1kgkz36rl4akv694yz65hr72q4kv4b3lxcbi65b3p' 
```
('sha256:0ip26j2h11n1kgkz36rl4akv694yz65hr72q4kv4b3lxcbi65b3p' means zero-length file)

It would be nice to create .drv files in a more redundant way excluding the possibility to end up with zero-length or truncated files in case of power outage (for example, create a `.drv.pending` file and then atomically rename to `.drv`)

**Comment by edolstra** [comment]
There is an option named `sync-before-registering` that ensures that store paths are written to disk before they're registered as valid.

**Comment by 7c6f434c** [comment]
>> There is an option named `sync-before-registering` that ensures that store paths are written to disk before they're registered as valid.
>Any reason why it is `false` by default?
>Performance gain seems negligible compared to the number of writes needed to build an average  derivation

Well, if these writes are all buffered, this syncing could be a large part of the build time for some derivations.

(But I agree that it is better to default to true here)

**Comment by edolstra** [comment]
I don't think it's neglible. `sync()` is a really slow system call on many filesystems. I've seen systems where it can take an unbounded amount of time (hours).

**Comment by bjornfor** [comment]
If it's just about .drv files(?) then one can perhaps use fsync()/fdatasync() instead?

**Comment by edolstra** [comment]
Yes (though `fsync()` is also expensive on some filesystems). But it wouldn't work for syncing derivation outputs since they can be directories.

**Comment by lheckemann** [comment]
I'm not sure I understand how the nix store becomes unusable through the presence of a few corrupted paths. That aside, would an alternative solution be to register paths before creating the file, and adding a "flushed" column or similar? An fsync could take place in the background (recursively in the case of directories) and be followed by updating the flushed column to ensure consistency and allow quick collection of the corrupted paths.
